### ISSUE TITLE
Permanent 301 redirect is not generated when product url changes on storeview scope

### ISSUE THREAD

**Issue by koenner01** [magento/magento2#13240]
When changing a product's url in the backend on storeview scope, there is no 301 rewrite being generated.

### Preconditions
1. PHP7 
2. CE 2.2.2 (we previously tested this issue too on 2.1.7+ and it existed there too)

### Steps to reproduce
1. Create a website and add multiple storeviews to it
2. Create a product, add it to your new website and then save it
3. Now change the store scope on the product detail in the backend to a storeview and change the product's url (**make sure the checkbox for 'Create Permanent Redirect for old URL' is checked!**) 

### Expected result
1.  A 301 redirect for the product's old url should've been created for choosen storeview

### Actual result
1. No 301 redirect is created

Weird thing is that if you update the url on 'all storeviews' scope it will create a 301 redirect for every storeview using the default value, so logic for the redirects must exist..
Not gonna bother going deeper into this issue as it's probably gonna be in the core depths of the magento (hell)

**Comment by magento-engcom-team** [comment]
@koenner01, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by diwipl** [comment]
I think I've found what the issue is.

https://github.com/magento/magento2/blob/2.2/app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php#L53-L58

Those lines remove old url_rewrite table entries for product for specified store view, but it is done before new redirects from old urls are generated. Because of that redirects cannot be generated because old urls does not exist anymore in database.

Removing those lines fixed the issue for me, but I cannot easily predict consequences of such modification in all parts of system. 

The reason why generation of redirects works for "All store views" is because `0` is passed as store id and url_rewrite table never contains entries with store_id equal to 0 (or at least in any Magento 2 database that I saw) so essentially nothing is removed and thus it is possible to generate redirects for old urls.

**Comment by magento-engcom-team** [comment]
Hi @koenner01. Thank you for your report. 
The issue has been fixed in magento/magento2#13567 by @adrien-louis-r in 2.2-develop branch
Related commit(s):
 - [1ef4dcc110062612e9971d7f8fb788bef213df3f](https://github.com/magento/magento2/commit/1ef4dcc110062612e9971d7f8fb788bef213df3f)
 - [864f371aed88884fb73082deb0c856819c43a41f](https://github.com/magento/magento2/commit/864f371aed88884fb73082deb0c856819c43a41f)

The fix will be available with the upcoming 2.2.4 release.

**Comment by magento-engcom-team** [comment]
Hi @koenner01. Thank you for your report. 
The issue has been fixed in magento/magento2#14252 by @rostyslav-hymon in 2.3-develop branch
Related commit(s):
 - [03e0a0b007c7d1f8a7f82d685a5e99aa1cd9e7c3](https://github.com/magento/magento2/commit/03e0a0b007c7d1f8a7f82d685a5e99aa1cd9e7c3)
 - [1322d8be3e64695effbfb525a22c3aed02587b1e](https://github.com/magento/magento2/commit/1322d8be3e64695effbfb525a22c3aed02587b1e)

The fix will be available with the upcoming 2.3.0 release.

**Comment by peterjaap** [comment]
@magento-engcom-team that is not the fix, that is just an integration test. Could you refer us to the actual commit that *fixes* it?

**Comment by hostep** [comment]
@peterjaap: indeed, but if you read the PR's description, you'll find a link to the fix: https://github.com/magento/magento2/commit/a0d3a637a08c2efa444a4fa259ac9d0d8100a113#diff-4e1fd4e484fc5142fc23f8ef0cde6718L53

A colleague of mine put this in a patch (see below) and reported it fixed the problem mentioned in this thread. However this not a fix for all the url rewrite issues when you have multiple storeviews, that is still more or less completely bugged (since Magento 2.0.0).

Patch for the `magento/module-catalog-url-rewrite` module:
```diff
diff --git a/Observer/ProductProcessUrlRewriteSavingObserver.php b/Observer/ProductProcessUrlRewriteSavingObserver.php
index ec20b04..8238978 100644
--- a/Observer/ProductProcessUrlRewriteSavingObserver.php
+++ b/Observer/ProductProcessUrlRewriteSavingObserver.php
@@ -50,12 +50,6 @@ class ProductProcessUrlRewriteSavingObserver implements ObserverInterface
             || $product->getIsChangedWebsites()
             || $product->dataHasChangedFor('visibility')
         ) {
-            $this->urlPersist->deleteByData([
-                UrlRewrite::ENTITY_ID => $product->getId(),
-                UrlRewrite::ENTITY_TYPE => ProductUrlRewriteGenerator::ENTITY_TYPE,
-                UrlRewrite::REDIRECT_TYPE => 0,
-                UrlRewrite::STORE_ID => $product->getStoreId()
-            ]);
 
             if ($product->isVisibleInSiteVisibility()) {
                 $this->urlPersist->replace($this->productUrlRewriteGenerator->generate($product));
diff --git a/Test/Unit/Observer/ProductProcessUrlRewriteSavingObserverTest.php b/Test/Unit/Observer/ProductProcessUrlRewriteSavingObserverTest.php
index 339ac02..95dfb65 100644
--- a/Test/Unit/Observer/ProductProcessUrlRewriteSavingObserverTest.php
+++ b/Test/Unit/Observer/ProductProcessUrlRewriteSavingObserverTest.php
@@ -205,13 +205,6 @@ class ProductProcessUrlRewriteSavingObserverTest extends \PHPUnit_Framework_Test
             ->method('getIsChangedCategories')
             ->will($this->returnValue($isChangedCategories));
 
-        $this->urlPersist->expects($this->exactly($expectedDeleteCount))->method('deleteByData')->with([
-            UrlRewrite::ENTITY_ID => $this->product->getId(),
-            UrlRewrite::ENTITY_TYPE => ProductUrlRewriteGenerator::ENTITY_TYPE,
-            UrlRewrite::REDIRECT_TYPE => 0,
-            UrlRewrite::STORE_ID => $this->product->getStoreId()
-        ]);
-
         $this->product->expects($this->any())
             ->method('isVisibleInSiteVisibility')
             ->will($this->returnValue($visibilityResult));
```

**Comment by peterjaap** [comment]
This fix is already in 2.2.3. What is exactly the change that will be in 2.3.0?

**Comment by hostep** [comment]
Hmm, I don't see the fix in 2.2.3, however, I can see it in 2.2.4.

Not sure if there will be a more elaborate fix in 2.3.0, but I doubt it. You can ignore those automated comments from @magento-engcom-team, they are very naive, in that they automatically put that message in all issues which get mentioned in a PR after it gets approved (which isn't always correct, but I guess it's better then not posting that message).
